core/include/traccc/edm/track_state.hpp

```diff
@@ -77,7 +77,11 @@ struct track_state {
                       "The measurement dimension should be 1 or 2");
 
         matrix_type<D, 1> ret;
-        matrix_operator().element(ret, 0, 0) = m_measurement.local[0];
+        if (m_measurement.subs.get_indices()[0] == 0)
```

Comment:
You should add a line of documentation that explains what this check is actually doing. Since I don't know this code, this line looks like checking some "magic number". :frowning:

Reply:
I had to rewind the logic from scratch...
What I think now is that we should not touch `track_state` at all

@shimasnd Have you checked the matrix elements of the projection matrix? https://github.com/acts-project/traccc/blob/fd145d8fa3b5dac40fa88e449d1d235f0dff6b85/core/include/traccc/fitting/kalman_filter/gain_matrix_updater.hpp#L74

`H` should be [0 1 0 0 0 0] for Annulus measurement so that  `H * predicted_vec` becaomes [e_bound_loc1], which is phi measurement.

Reply:
No I think actually this PR is doing the right thing as the following line should bring the `e_bound_loc1` component when `subs.get_indices()[0] == 1`.

https://github.com/acts-project/traccc/blob/fd145d8fa3b5dac40fa88e449d1d235f0dff6b85/core/include/traccc/fitting/kalman_filter/gain_matrix_updater.hpp#L77-L78

I guess we should be generous about using the magic number here but we can be more explicit..

```c++
if (m_measurement.subs.get_indices()[0] == 0){
            matrix_operator().element(ret, 0, 0) = m_measurement.local[0];
            if constexpr (D == 2u) {
                 matrix_operator().element(ret, 1, 0) = m_measurement.local[1];
            }
}
else if (m_measurement.subs.get_indices()[0] == 1){
            matrix_operator().element(ret, 0, 0) = m_measurement.local[1];
            if constexpr (D == 2u) {
                 matrix_operator().element(ret, 1, 0) = m_measurement.local[0];
            }
}
```

And the same goes for variances.